tf2onnx/tfonnx.py

```diff
@@ -1616,6 +1616,144 @@ def where_op(ctx, node, name, args):
+    o = sigmoid(cs * wco + o)
+    co = tanh(cs)
+    h = co .* o
+    ```
+    """
+    nodes = []
+    x, cs_prev, h_prev, w, wci, wcf, wco, b = node.input
+    forget_bias = float(node.get_attr("forget_bias").f)
+    cell_clip = float(node.get_attr("cell_clip").f)
+    use_peephole = bool(node.get_attr("use_peephole").i)
+
+    def make_sigmoid(i, w, b):
+        i_w_node = ctx.make_node("Mul", [i, w])
+        i_w_b_node = ctx.make_node("Add", [i_w_node.output[0], b])
+        output_node = ctx.make_node("Sigmoid", [i_w_b_node.output[0]])
+        nodes.extend([i_w_node, i_w_b_node, output_node])
+        return output_node.output[0]
+
+    # xh = [x, h]
+    xh_node = ctx.make_node("Concat", [x, h_prev], attr={"axis": 1})
+
+    # i, ci, f, o = xh * w + b
+    xh_w_node = ctx.make_node("MatMul", [xh_node.output[0], w])
+    w_shape = ctx.get_shape(w)
+    if len(w_shape) != 2 or w_shape[1] % 4 != 0:
+        raise RuntimeError("shape of W of LSTMBlockCell {} should be times of 4".format(name))
+    merged_output_node = ctx.make_node("Add", [xh_w_node.output[0], b])
+    w_last_dim = int(w_shape[1] / 4)
+    split = [w_last_dim] * 4
+    split_output_node = ctx.make_node(
+        "Split", [merged_output_node.output[0]],
+        attr={"axis": 1, "split": split},
+        output_count=4
+    )
+    i, ci, f, o = split_output_node.output
+
+    # f = f + forget_bias
+    forget_bias_const = ctx.make_const(
+        utils.make_name("{}__forget_bias".format(name)),
+        np.array(forget_bias, dtype=np.float32)
+    )
+    f_node = ctx.make_node("Add", [f, forget_bias_const.output[0]])
+
+    if not use_peephole:
+        zeros_const = ctx.make_const(
+            utils.make_name("{}__zeros_const".format(name)),
+            np.zeros([w_last_dim], dtype=np.float32)
+        )
+        nodes.append(zeros_const)
+        wci = zeros_const.output[0]
+        wcf = zeros_const.output[0]
+        wco = zeros_const.output[0]
+
+    # i = sigmoid(cs_prev .* wci + i)
+    i = make_sigmoid(cs_prev, wci, i)
+    # f = sigmoid(cs_prev .* wcf + f)
+    f = make_sigmoid(cs_prev, wcf, f_node.output[0])
+    # ci = Tanh(ci)
+    ci_node = ctx.make_node("Tanh", [ci])
+    # cs = ci .* i + f .* cs_prev
+    ci_i_node = ctx.make_node("Mul", [ci_node.output[0], i])
+    cs_prev_f_node = ctx.make_node("Mul", [cs_prev, f])
+    cs_node = ctx.make_node("Add", [ci_i_node.output[0], cs_prev_f_node.output[0]])
+    cs = cs_node.output[0]
+    # cs = clip(cs)
+    if cell_clip > 0:
+        cs_clip_node = ctx.make_node("Clip", [cs], attr={"max": cell_clip, "min": -cell_clip})
+        nodes.append(cs_clip_node)
+        cs = cs_clip_node.output[0]
+    # o = cs * wco + o
+    o = make_sigmoid(cs, wco, o)
+    # co = Tanh(cs)
+    co_node = ctx.make_node("Tanh", [cs])
+    # h = co .* o
+    h_node = ctx.make_node("Mul", [co_node.output[0], o])
+
+    def replace_output(old_output, new_output):
+        ctx.replace_all_inputs(ctx.get_nodes(), old_output, new_output)
```

Comment:
be noted, with my PR merged, ctx.get_nodes() already get all update nodes, including the one you created. Be careful there is node having input_id == output_id after replace_all_inputs.

Reply:
maybe add a warning for such case? at least for now.